src/torchcodec/decoders/_core/CudaDevice.cpp

```diff
@@ -10,13 +51,83 @@ void throwErrorIfNonCudaDevice(const torch::Device& device) {
     throw std::runtime_error("Unsupported device: " + device.str());
   }
 }
+} // namespace
 
-void initializeDeviceContext(const torch::Device& device) {
+void initializeDeviceContext(
+    const torch::Device& device,
+    AVCodecContext* codecContext) {
   throwErrorIfNonCudaDevice(device);
-  // TODO: https://github.com/pytorch/torchcodec/issues/238: Implement CUDA
-  // device.
-  throw std::runtime_error(
-      "CUDA device is unimplemented. Follow this issue for tracking progress: https://github.com/pytorch/torchcodec/issues/238");
+  // This is a dummy tensor to initialize the cuda context.
+  torch::Tensor dummyTensorForCudaInitialization = torch::empty(
+      {1}, torch::TensorOptions().dtype(torch::kUInt8).device(device));
+  codecContext->hw_device_ctx = av_buffer_ref(getCudaContext(device));
+  return;
+}
+
+void convertAVFrameToDecodedOutputOnDevice(
+    const torch::Device& device,
+    const VideoDecoder::VideoStreamDecoderOptions& options,
+    AVCodecContext* codecContext,
+    VideoDecoder::RawDecodedOutput& rawOutput,
+    VideoDecoder::DecodedOutput& output) {
+  AVFrame* src = rawOutput.frame.get();
+
+  TORCH_CHECK(
+      src->format == AV_PIX_FMT_CUDA,
+      "Expected format to be AV_PIX_FMT_CUDA, got " +
+          std::string(av_get_pix_fmt_name((AVPixelFormat)src->format)));
+  int width = options.width.value_or(codecContext->width);
+  int height = options.height.value_or(codecContext->height);
+  NppStatus status;
+  NppiSize oSizeROI;
+  oSizeROI.width = width;
+  oSizeROI.height = height;
```

Comment:
Nit: we can do this cleaner with:
```suggestion
  NppiSize oSizeROI = {width, height];
```
A theme to my nits is that if possible, I prefer to initialize and declare values at the same time. :)